v2:  Expose TLS placeholders

moorereason opened the pull request:
This PR will fix #2505 and #2977.

I updated the const names in `caddyhttp/replacer.go` which creates a little diff churn.  The naming was a little inconsistent.  I couldn't help myself.  :smiley:

Is `{http.request.tls.client.cert}` useful as `string(cert.Raw)`?  I ported the v1 behavior, but it seems like that byte slice should be encoded to printable text.  I'm not sure what it's used for.

Second commit may be off-topic, but it's trivial.  Let me know if I need to revert it.

I'm new to Caddy, so suggest any changes.

mholt commented:
Woah, this looks like a great PR. I can't wait to review it! Lemme get beta 13 out the door first and I'll head back to this. Thanks!

mholt commented on modules/caddytls/values.go:
```diff
@@ -127,9 +128,35 @@ var SupportedProtocols = map[string]uint16{
 	"tls1.3": tls.VersionTLS13,
 }
 
+// UnsupportedProtocols is a map of unsupported protocols.
```
It might be worth noting in the godoc that this is just for logging purposes, not enforcement.

moorereason replied:
Pushed a commit to update the godoc and not export `unsupportedProtocols`.  Let me know if you agree with the change.

mholt commented on modules/caddyhttp/replacer.go:
```diff
@@ -208,12 +217,79 @@ func addHTTPVarsToReplacer(repl *caddy.Replacer, req *http.Request, w http.Respo
 	repl.Map(httpVars)
 }
 
+func getReqTLSReplacement(req *http.Request, key *string) (string, bool) {
```
Why is `key` a pointer? I believe empty string is also invalid, so there is probably no need for a nil key, right?

moorereason replied:
I'm was trying to reduce memory allocations.  Let me know if you would rather we just pass a string.

mholt replied:
Let's just pass a string for now; an optimization like this should be profiled first.

moorereason replied:
Fixed.

mholt commented on modules/caddyhttp/replacer.go:
```diff
@@ -208,12 +217,79 @@ func addHTTPVarsToReplacer(repl *caddy.Replacer, req *http.Request, w http.Respo
+	if len(cs.PeerCertificates) == 0 {
+		return nil
+	}
+	return cs.PeerCertificates[0]
```
I can't remember, if the peer presents a certificate chain (bundle), will this be the leaf certificate or the CA? Which one will we want? Does it depend? We should consider these questions.

moorereason replied:
Good question.  I'll try to find an answer.  I'm pretty sure I mirrored the behavior of v1 for this.

moorereason replied:
Here's what I see in the stdlib:

The godocs for [`tls.ConnectionState`](https://golang.org/pkg/crypto/tls/#ConnectionState) `PeerCertificates` are not explicit about the leaf cert being first.  There's a hint in [`tls.Certificate`](https://golang.org/pkg/crypto/tls/#Certificate), which says "A Certificate is a chain of one or more certificates, leaf first," but `PeerCertificates` is not of type `tls.Certificate`.

Browsing through [`crypto/tls/handshake_server.go#L545`](https://github.com/golang/go/blob/88e564edb13f1596c12ad16d5fd3c7ac7deac855/src/crypto/tls/handshake_server.go#L545), it certainly appears that the first certificate is assumed to be the leaf.

mholt replied:
Indeed, it also seems to be the case here: https://github.com/golang/go/blob/33d20760a23acffd3b200b922c5687ed2aeb74cc/src/crypto/tls/conn.go#L1419

It's not strictly documented anywhere, but let's run with that, I guess.

mholt commented on modules/caddytls/cipher_suites.go:
```diff
@@ -0,0 +1,107 @@
+// Derived from https://github.com/golang/go/commit/0ee22d9 which should be
+// included in Go 1.14.  Once Go 1.14 is released, we can remove this file.
```
Go 1.14 is in RC, and I think we won't get around to merging this until Go 1.14 is released, probably within the next couple weeks. So, just keep in mind, I think we will be removing this file before the merge.

moorereason replied:
:+1: I'm in no rush.  Happy to wait for Go 1.14 to land.

mholt commented:
@moorereason It's looking good! After Go 1.14 is released, let's remove that redundant file and then I'll approve this for merging if there are no conflicts.

mholt commented:
> Is {http.request.tls.client.cert} useful as string(cert.Raw)? I ported the v1 behavior, but it seems like that byte slice should be encoded to printable text.

I just saw this, somehow. So actually, I think you make a good point... let's remove that particular placeholder and see what the actual needs are in Caddy 2 first. If we get requests for it, we can add it (maybe base64-encoded?) but no need to have it for right now.

moorereason commented:
> > Is {http.request.tls.client.cert} useful as string(cert.Raw)? I ported the v1 behavior, but it seems like that byte slice should be encoded to printable text.
> 
> I just saw this, somehow. So actually, I think you make a good point... let's remove that particular placeholder and see what the actual needs are in Caddy 2 first. If we get requests for it, we can add it (maybe base64-encoded?) but no need to have it for right now.

Removed in latest commit.  Also rebased to fix a merge conflict.

I'll be back once Go 1.14 is released.  👀 👍

mholt commented:
@moorereason Go 1.14 was released today, shall we finish this up!? :+1:

moorereason commented:
Pushed a commit for Go 1.14 support.  🚀

Mohammed90 commented:
If this PR needs Go 1.14, then the go.mod file needs to be updated too.

mholt commented:
Bumped to Go 1.14 in https://github.com/caddyserver/caddy/commit/78760c0ddcd890702977c876fdb126026e13f937
